pkg/virtualservice/resource_manager.go

```diff
@@ -0,0 +1,345 @@
+		return sdkVS, nil
+	}
+	if !m.isSDKVirtualServiceControlledByCRDVirtualService(ctx, sdkVS, vs) {
+		m.log.V(2).Info("skip virtualService update since it's not controlled",
+			"virtualService", k8s.NamespacedName(vs),
+			"virtualServiceARN", aws.StringValue(sdkVS.Metadata.Arn),
+		)
+		return sdkVS, nil
+	}
+
+	diff := cmp.Diff(desiredSDKVSSpec, actualSDKVSSpec, opts)
+	m.log.V(2).Info("virtualServiceSpec changed",
+		"virtualService", k8s.NamespacedName(vs),
+		"actualSDKVRSpec", actualSDKVSSpec,
+		"desiredSDKVRSpec", desiredSDKVSSpec,
+		"diff", diff,
+	)
+	resp, err := m.appMeshSDK.UpdateVirtualServiceWithContext(ctx, &appmeshsdk.UpdateVirtualServiceInput{
+		MeshName:           sdkVS.MeshName,
+		MeshOwner:          sdkVS.Metadata.MeshOwner,
+		VirtualServiceName: sdkVS.VirtualServiceName,
+		Spec:               desiredSDKVSSpec,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return resp.VirtualService, nil
+}
+
+func (m *defaultResourceManager) deleteSDKVirtualService(ctx context.Context, sdkVS *appmeshsdk.VirtualServiceData, vs *appmesh.VirtualService) error {
+	if !m.isSDKVirtualServiceOwnedByCRDVirtualService(ctx, sdkVS, vs) {
+		m.log.V(2).Info("skip virtualService deletion since its not owned",
+			"virtualService", k8s.NamespacedName(vs),
+			"virtualServiceARN", aws.StringValue(sdkVS.Metadata.Arn),
+		)
+		return nil
+	}
+	_, err := m.appMeshSDK.DeleteVirtualServiceWithContext(ctx, &appmeshsdk.DeleteVirtualServiceInput{
+		MeshName:           sdkVS.MeshName,
+		MeshOwner:          sdkVS.Metadata.MeshOwner,
+		VirtualServiceName: sdkVS.VirtualServiceName,
+	})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (m *defaultResourceManager) updateCRDVirtualService(ctx context.Context, vs *appmesh.VirtualService, sdkVS *appmeshsdk.VirtualServiceData) error {
+	oldVS := vs.DeepCopy()
+	needsUpdate := false
+	if aws.StringValue(vs.Status.VirtualServiceARN) != aws.StringValue(sdkVS.Metadata.Arn) {
+		vs.Status.VirtualServiceARN = sdkVS.Metadata.Arn
+		needsUpdate = true
+	}
+	vsActiveConditionStatus := corev1.ConditionFalse
+	if sdkVS.Status != nil && aws.StringValue(sdkVS.Status.Status) == appmeshsdk.VirtualServiceStatusCodeActive {
+		vsActiveConditionStatus = corev1.ConditionTrue
+	}
+	if updateCondition(vs, appmesh.VirtualServiceActive, vsActiveConditionStatus, nil, nil) {
+		needsUpdate = true
+	}
+
+	if !needsUpdate {
+		return nil
+	}
+	return m.k8sClient.Patch(ctx, vs, client.MergeFrom(oldVS))
+}
+
+// isSDKVirtualServiceControlledByCRDVirtualService checks whether an AppMesh VirtualService is controlled by CRD VirtualService.
+// if it's controlled, CRD VirtualService update is responsible for update AppMesh VirtualService.
+func (m *defaultResourceManager) isSDKVirtualServiceControlledByCRDVirtualService(ctx context.Context, sdkVS *appmeshsdk.VirtualServiceData, vs *appmesh.VirtualService) bool {
+	if aws.StringValue(sdkVS.Metadata.ResourceOwner) != m.accountID {
+		return false
+	}
+	return true
```

Comment:
Why not just:
```go
return aws.StringValue(sdkVS.Metadata.ResourceOwner) == m.accountID
```

Reply:
the original purpose is there might be additional logic add here, so i used this pattern :D

Reply:
I could see that for `isSDKVirtualServiceOwnedByCRDVirtualService()` where the TODO is, but not really in this function. :smile:

Reply:
haha, changed